### ISSUE TITLE
Generic SASL library

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2015]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2015|
|Type|enhancement|
|Created|2006-08-23 15:05:53Z|
|Branch|https://github.com/twisted/twisted/tree/generic-sasl-2015-2|

Cred is already general enough for most pieces of the authentication/authorization process to be re-used.  However, various concrete SASL mechanisms are being re-implemented all over the place.

Cred should at least come with an implementation of:

  LOGIN
  PLAIN
  CRAM-MD5
  DIGEST-MD5

Implementations of which can already be found scattered around the Twisted codebase.

Attachments:

* [sasl.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9cd/9cdda67ded3f25811728276cefa76b80913b4c54/84c3cdfcb7b83979e1a0e842668790d3a9465a7c.patch) (75965 bytes) - added by antoine on 2007-03-14 15:27:35Z - 
* [concrete-sasl.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9cd/9cdda67ded3f25811728276cefa76b80913b4c54/42827894f2c45ffdc3f6aa69a992e92bbd899cd6.patch) (63946 bytes) - added by antoine on 2007-03-14 15:29:20Z - 
* [sasl-doc-fixes.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9cd/9cdda67ded3f25811728276cefa76b80913b4c54/5a6a33f28870c62dfc473a38ffbfc4551b6d9432.patch) (39131 bytes) - added by antoine on 2007-05-30 12:44:47Z - 
* [sasl-unicode-strict.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9cd/9cdda67ded3f25811728276cefa76b80913b4c54/cea83f414e639dd55205ff030044aa2a94c48511.patch) (17500 bytes) - added by antoine on 2007-05-30 13:51:04Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2015 2015
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__generic_sasl_2015_2 generic-sasl-2015-2
branch_author__antoine__therve antoine, therve
status__new new
resolution__None None
component__core core
keywords__ 
time__1156345553000000 1156345553000000
changetime__1302120131000000 1302120131000000
version__None None
owner__ 
cc__ralphm cc__antoine cc__therve cc__dreid cc__njoyce cc__exarkun cc__khorn cc__thijs cc__whitelynx
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm commented|
|-|-|

Most notably, there should be implementations of SASL authentication mechanisms for both the client and server side. They should be protocol agnostic, so they can be used with IMAP, XMPP, LDAP, etc. Also note that there is an updated SASL RFC now: RFC 4422.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

(In [18775]) Move public Jabber interface definitions to one module.

Author: ralphm
Reviewer: jml
Fixes [#2177](https://github.com/twisted/twisted/issues/2177)

The interface definitions in sasl_mechanisms are left there, because this
module should really move to or be merged with a more generic SASL
implementation (see [#2015](https://github.com/twisted/twisted/issues/2015)).

The interface definitions in xmpp_stringprep are left there, because they
are considered private and for documentation only.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_4"></a>|@pitrou commented|
|-|-|

I'd like to try to work on this. My main focus is the Digest mechanism, which needs to support both real SASL digest and legacy HTTP digest (the latter being also used in SIP).

I propose to add a twisted.cred.sasl module for generic SASL interfaces, and a twisted.cred.digest module for Digest implementation.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm commented|
|-|-|

Sounds great. In the stuff I did in `twisted.words.protocols.jabber.sasl_mechanisms`, I tried to abstract away from the SASL profiles. All XMPP specific bits are in `twisted.words.protocols.jabber.sasl`. Maybe that's a good start. Note that I only did the client side of things for now.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_6"></a>|@pitrou commented|
|-|-|

Hi,

I've just attached a draft implementation.
It adds the following files:
 * twisted.cred.sasl contains generic SASL interfaces, error classes et al.
 * twisted.cred.digest contains Digest implementations supporting the three RFCs (2831, 2617, 2069)
 * twisted.test.test_digest contains tests for the Digest implementations

Things to look at in particular:
 * the interfaces in twisted.cred.sasl: ISASLChallenger, ISASLResponder
 * the high-level classes in twisted.cred.digest: SASLDigestChallenger, SASLDigestResponder, HTTPDigestChallenger, HTTPDigestResponder

Some pending issues:
 * rspauth with HTTP Digest and qop=auth-int is probably wrong (I don't know which entity-body should be used: the one from the response or the one from the final challenge?)
 * the nonce-count is not verified against the nonce by (SASL|HTTP)DigestChallenger, yet.

Unimplemented stuff:
 * integrity checking for further messages (with qop=auth-int)
 * qop=auth-conf
 * nextnonce and company in the final challenge for HTTP Digest auth

Hopefully this code is still a superset of existing SASL implementations in Twisted (I've only looked at Web2, Jabber and SIP to be honest).

I'm looking for feedback.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_7"></a>|@pitrou commented|
|-|-|

I meant a superset of existing SASL _Digest_ implementations of course.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_8"></a>|@pitrou commented|
|-|-|

I updated the patch to make it Python 2.3-compatible.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm commented|
|-|-|

At first glance, it seems strange to me that you yank the HTTP peculiarities all the way up to the generic SASL interface, while HTTP doesn't actually _do_ SASL. Unlike SMTP, XMPP and LDAP.

To me, it would be better to make the SASL Challenger and Responder interfaces (as you call them) pure SASL, that deal with different authentication mechanisms, and make a separate set of Challenger and Responder for the HTTP Digest case. You can still reuse the common bits in `digest.py`, while having cleaner interfaces for the SASL side.

I need to look into your large patch in more detail, still.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_10"></a>|@pitrou commented|
|-|-|

Hmm, yes you are probably right about the interfaces. I'm gonna solve this problem.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_12"></a>|@pitrou commented|
|-|-|

Updated patch to make SASL interfaces pure SASL.
HTTP Digest and SASL Digest implementations still derive from the same base class, but only the derived classes for SASL implement the interfaces. I had to do a bit of docstring duplication.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_13"></a>|@pitrou commented|
|-|-|

By the way, don't bother about the privateKey stuff in BaseDigestChallenger (which was taken from web2.auth.digest). I'm gonna replace it with a nonce table which will also allow checking the nonce-count values (protecting against replays).

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_14"></a>|@pitrou commented|
|-|-|

Update the patch with new version.
The privateKey generation is removed and replaced with a dict of generated nonces. It handles expiration of nonces as well as protection against replays (by checking and remembering the nc value).

Can someone have a look at the patch? I don't see any big issues with it myself.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

A couple minor points from a really superficial look at the patch:

  1. zope.interface.Interface methods shouldn't be defined with ```self``` parameter
  2. some tests are defined as testFoo rather than test_foo and are missing docstrings
  3. There is some random number code in twisted.names and twisted.conch, it might be good to factor that out into twisted.python and re-use it here

It'd also be pretty useful to see this code being used to replace some existing digest code in Twisted.

I'm also not sure I see support for multiple roundtrips in the ISASLChallenger interface, but I might have just not looked hard enough.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_16"></a>|@pitrou commented|
|-|-|

Hi,

> A couple minor points from a really superficial look at the patch:
>
>   1. zope.interface.Interface methods shouldn't be defined with ```self``` parameter

Ok, will be fixed in the next upload.

>   2. some tests are defined as testFoo rather than test_foo and are missing docstrings

Ditto.

>   3. There is some random number code in twisted.names and twisted.conch, it might be good to factor that out into twisted.python and re-use it here

Agreed. I didn't want to make the patch too hard to understand so I avoided touching existing files, but it will be definitely useful to factor out the secure random code in twisted.python once the rest is settled.

> It'd also be pretty useful to see this code being used to replace some existing digest code in Twisted.

I'm really not familiar enough with the relevant subpackages to be sure there aren't any subtle points that I'm missing. I'd prefer this part to be done by the respective maintainers honestly.

> I'm also not sure I see support for multiple roundtrips in the ISASLChallenger interface, but I might have just not looked hard enough.

I'm gonna add some tests for this in the next upload.

Thanks a lot for your feedback

Antoine.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_17"></a>|@pitrou commented|
|-|-|

New patch with fixed test naming, fixed method signature in Interfaces, added docstrings in tests, and added tests for multiple roundtrips in (SASL|HTTP)DigestChallenger (both full roundtrips and partial roundtrips (i.e. multiple responses produced from a single challenge) are tested for).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_18"></a>|@ralphm commented|
|-|-|

Personally, I would not add this interface support for multiple roundtrips. Each authentication mechanism may have its own requirements in how different challenges are to be processed, and may or may not need to keep state. It comes down to this, I suppose: how many instances will be needed for a typical use? One responder/challenger per connection?

For the `ISASLResponder` interface, I see the `uri` parameter. This hardly seems necessary to me. I didn't need one for the stuff I did in the Jabber SASL implementation. Also, looking at the tests, the values used there for this parameter don't look like URIs.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_19"></a>|@pitrou commented|
|-|-|

Added tests for non-ASCII usernames.
Added support for realmless challenges.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_20"></a>|@pitrou commented|
|-|-|

Replying to [ralphm](#note_18):
> Personally, I would not add this interface support for multiple roundtrips. Each authentication mechanism may have its own requirements in how different challenges are to be processed, and may or may not need to keep state.

Well, unless I'm missing something, supporting multiple roundtrips does not change anything to the interface definition.

> It comes down to this, I suppose: how many instances will be needed for a typical use? One responder/challenger per connection?

For connected protocols, yes.

> For the `ISASLResponder` interface, I see the `uri` parameter. This hardly seems necessary to me. I didn't need one for the stuff I did in the Jabber SASL implementation.

Well, you have a (serv_type, host, serv_name) tuple instead :-)
As far as the SASL responder is concerned, the uri is an opaque value. It's up to the protocol implementation to generate it.

The problem is that the SASL spec itself is too vague, it claims to specify an abstraction without giving an explicit API. So I had to work out an API myself.

> Also, looking at the tests, the values used there for this parameter don't look like URIs.

Some of them are "digest URIs" as specified in SASL digest, some of them are "URIs" as specified in HTTP digest. Since they are treated as opaque strings it doesn't matter really.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_21"></a>|@pitrou commented|
|-|-|

New patch: made it actually compliant with the username and password encoding requirements in RFC 2831. Distinguished the latin and non-latin cases in the tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

Replying to [antoine](#note_16):

> > It'd also be pretty useful to see this code being used to replace some existing digest code in Twisted.
>
> I'm really not familiar enough with the relevant subpackages to be sure there aren't any subtle points that I'm missing. I'd prefer this part to be done by the respective maintainers honestly.

Thanks for the patch, of course, but - a first step to writing a library which unifies things in multiple packages is to become familiar with those things in multiple packages :).

The purpose of this ticket is to produce a generalized implementation of SASL methods which are used by all of the places _within_ Twisted that already implement SASL methods, not to provide yet another slightly incompatible implementation of digest auth that everyone can use.  Although the implementation here looks relatively clean and better documented than some of the others we currently have, it really just makes the existing problem (too many SASL method implementations) worse, not better.

In other words, a key requirement of any patch getting accepted for this feature is implementing at least _two_ protocols' SASL support in terms of the new module.  That's the only reasonable way to demonstrate it is plausible that it really is a "generic SASL library", and can one day replace them all.

You don't have to do this _yourself_, mind you, merely champion the patch to enough people that it gets done somehow.  You could try for the package maintainers, but I think you'll find they are unfortunately busy with other stuff...

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_23"></a>|@pitrou commented|
|-|-|

Ok, so here is a patch for twisted.protocols.sip and its associated tests. I called the patch concrete-sasl.patch so that the main patch does not get more difficult to read because of this.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_24"></a>|@pitrou commented|
|-|-|

Updated concrete-sasl.patch. It now contains proof-of-concept adaptations for SIP (twisted.protocols.sip) and HTTP (twisted.web2.auth.digest).
:-)

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_25"></a>|@pitrou commented|
|-|-|

New sasl.patch:
 * added implementation of SASL PLAIN (RFC 4616)
 * added support for authzid (with an IAuthzID interface)

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_26"></a>|@pitrou commented|
|-|-|

New concrete-sasl.patch. Now provides proof-of-concept adaptations for:
 * SIP (twisted.protocols.sip)
 * HTTP (twisted.web2.auth.digest)
 * XMPP (twisted.words.protocols.jabber.sasl_mechanisms)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_28"></a>|@therve removed owner|
|-|-|

I went on and check the initial patch in generic-sasl-2015. For now the concrete implementations are left out.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_29"></a>|@ralphm commented|
|-|-|

Replying to [antoine](#note_20):
> Replying to [ralphm](#note_18):
> > Personally, I would not add this interface support for multiple roundtrips. Each authentication mechanism may have its own requirements in how different challenges are to be processed, and may or may not need to keep state.
>
> Well, unless I'm missing something, supporting multiple roundtrips does not change anything to the interface definition.

Sure it does. You changed what is returned by the methods!

> > It comes down to this, I suppose: how many instances will be needed for a typical use? One responder/challenger per connection?
>
> For connected protocols, yes.

Ok

> > For the `ISASLResponder` interface, I see the `uri` parameter. This hardly seems necessary to me. I didn't need one for the stuff I did in the Jabber SASL implementation.
>
> Well, you have a (serv_type, host, serv_name) tuple instead :-)
> As far as the SASL responder is concerned, the uri is an opaque value. It's up to the protocol implementation to generate it.
>
> The problem is that the SASL spec itself is too vague, it claims to specify an abstraction without giving an explicit API. So I had to work out an API myself.

Sure. But this `uri` parameter is by no means a generic parameter for all possible mechanisms that do or do not yet exist. Yes, I did use (`serv_type`, `host`, `serv_name`) tuple, but this is specific for the `DIGEST-MD5` mechanism. Thus, they are instance variables of the `Digest` mechanism object.

The reason I did not use a `uri` parameter to the `__init__` method of `Digest` is that sometimes the `serv_name` is not used (`None`), like for XMPP. This way you can keep the actual construction of the `digest-uri` field generic. But this might make it a bit more difficult for making it usable for HTTP Digest, so I wouldn't mind just passing the `digest-uri` as a parameter to `__init__` instead of the tuple.

But, again, I don't think you should have to pass it to `getInitialResponse` and `getResponse` as a parameter.

So, if you're ok with that, I'll modify this in the branch therve started. There's also various comments that need updates since you removed parameters earlier.

> > Also, looking at the tests, the values used there for this parameter don't look like URIs.
>
> Some of them are "digest URIs" as specified in SASL digest, some of them are "URIs" as specified in HTTP digest. Since they are treated as opaque strings it doesn't matter really.

Hah, my bad. Looking at RFC 2831 again, please ignore that comment.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_30"></a>|@ralphm commented|
|-|-|

Oh, by the way. You are reading RFC 4422 instead of RFC 2222, right?

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_31"></a>|@pitrou commented|
|-|-|

Hi again !

Replying to [ralphm](#note_29):
> > Well, unless I'm missing something, supporting multiple roundtrips does not change anything to the interface definition.
>
> Sure it does. You changed what is returned by the methods!

I'm not sure I understand you. You are talking about the "challenge type" returned as part of ISASLResponder.getResponse, right ?
Then not supporting multiple roundtrips doesn't change anything, since there must still be a way to indicate that the challenge was a successful one (e.g. "rspauth").

Again, I'm not sure how this is important, since the existing implementations don't have common interfaces right now.

> Sure. But this `uri` parameter is by no means a generic parameter for all possible mechanisms that do or do not yet exist.

Agreed. It is needed however for multiple responses to a same challenge with Digest.
It is quite mandatory for HTTP. I don't know how useful it is with SASL Digest, but RFC 2831 doesn't deprecate the "nonce-count" and "stale" parameters, so I suppose multiple responses to a given challenge should still be supported.

What do you think?

> Oh, by the way. You are reading RFC 4422 instead of RFC 2222, right?

Yes, but both seem unfortunately too vague to me, that's why I'm trying to be pragmatic about the solution.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_32"></a>|@ralphm commented|
|-|-|

Replying to [antoine](#note_31):
> Hi again !
>
> Replying to [ralphm](#note_29):
> > > Well, unless I'm missing something, supporting multiple roundtrips does not change anything to the interface definition.
> >
> > Sure it does. You changed what is returned by the methods!
>
> I'm not sure I understand you. You are talking about the "challenge type" returned as part of ISASLResponder.getResponse, right ?
> Then not supporting multiple roundtrips doesn't change anything, since there must still be a way to indicate that the challenge was a successful one (e.g. "rspauth").

Well, what you mean is that if the SASL-using protocol (like IMAP) doesn't support sending along additional data on success, the last challenge by the server carries this additional data and must be responded to with an empty response. An alternative to returning an object that may indicate such additional data, you could set an attribute on the SASL mechanism instance that can be read when the protocol handler receives the success indication. This is how I planned to support this in my abstraction.

> Again, I'm not sure how this is important, since the existing implementations don't have common interfaces right now.

Sure. I am just trying to get a nice clean interface here.

> > Sure. But this `uri` parameter is by no means a generic parameter for all possible mechanisms that do or do not yet exist.
>
> Agreed. It is needed however for multiple responses to a same challenge with Digest.
> It is quite mandatory for HTTP.

Ah, but we are building a generic SASL interface plus mechanism implementations. HTTP doesn't do SASL and you can that differently
for the HTTP Digest stuff like with the other parameters you removed earlier. As you mentioned the `uri` parameter is obviously specific to the `DIGEST-MD5` mechanism, so it should not be in the generic interface.

> I don't know how useful it is with SASL Digest, but RFC 2831 doesn't deprecate the "nonce-count" and "stale" parameters, so I suppose multiple responses to a given challenge should still be supported.

I don't understand what you are saying here.

> > Oh, by the way. You are reading RFC 4422 instead of RFC 2222, right?
>
> Yes, but both seem unfortunately too vague to me, that's why I'm trying to be pragmatic about the solution.

Ah, and I am trying to do it well from the start. I have studied SASL in detail when implementing it for the XMPP stuff. I'm sure we'll end with a nice pragmatic but also clean solution.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_33"></a>|@ralphm commented|
|-|-|

This [message](http://www.imc.org/ietf-sasl/mail-archive/msg02799.html) on the IETF SASL mailing list might be interesting, by the way.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_34"></a>|@pitrou commented|
|-|-|

Replying to [ralphm](#note_32):
> An alternative to returning an object that may indicate such additional data, you could set an attribute on the SASL mechanism instance that can be read when the protocol handler receives the success indication.

But it makes it more difficult, or at least error-prone, to create several responses with the same responder. Which is useful for non-connected protocols.

> > I don't know how useful it is with SASL Digest, but RFC 2831 doesn't deprecate the "nonce-count" and "stale" parameters, so I suppose multiple responses to a given challenge should still be supported.
>
> I don't understand what you are saying here.

What I'm saying is that if the RFC 2831 authors meant that only a single response should be generated for each Digest challenge, then they would have deprecated the nonce-count and stale parameters, or discouraged against their use, since they wouldn't be useful anymore. There's nothing inside RFC 2831 that says generating several responses from the same challenge is not ok. There's nothing suggesting those responses must have the same digest-uri.

If the above is true, then the digest-uri parameter has to stay part of the getInitialResponse/getResponse signature.

Perhaps someone else familiar with SASL, Digest and friends can share his opinion here.

> This message on the IETF SASL mailing list might be interesting, by the way.

Interesting indeed. But this is very forward-looking...
Digest will be around for years, and the other common SASL mechanisms (CRAM-MD5, PLAIN) are more insecure.

(from a bit of reading [2831bis](http://tools.ietf.org/html/draft-ietf-sasl-rfc2831bis), it seems to add even more complexity... let's hope we don't have to support it)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_35"></a>|@ralphm commented|
|-|-|

Replying to [antoine](#note_34):
> Replying to [ralphm](#note_32):
> > An alternative to returning an object that may indicate such additional data, you could set an attribute on the SASL mechanism instance that can be read when the protocol handler receives the success indication.
>
> But it makes it more difficult, or at least error-prone, to create several responses with the same responder. Which is useful for non-connected protocols.
>
> > > I don't know how useful it is with SASL Digest, but RFC 2831 doesn't deprecate the "nonce-count" and "stale" parameters, so I suppose multiple responses to a given challenge should still be supported.
> >
> > I don't understand what you are saying here.
>
> What I'm saying is that if the RFC 2831 authors meant that only a single response should be generated for each Digest challenge, then they would have deprecated the nonce-count and stale parameters, or discouraged against their use, since they wouldn't be useful anymore. There's nothing inside RFC 2831 that says generating several responses from the same challenge is not ok. There's nothing suggesting those responses must have the same digest-uri.
>
> If the above is true, then the digest-uri parameter has to stay part of the getInitialResponse/getResponse signature.
>
> Perhaps someone else familiar with SASL, Digest and friends can share his opinion here.

Apparently you are talking about 'Subsequent Authentication'. Could you please use the terminology as used in the specification, because otherwise I get very confused.

Ok. I see that you indeed need the value of `digest-uri` for subsequent authentication, but that is _still_ specific to the DIGEST-MD5 authentication mechanism, and I think it should not be present in an interface to multiple different authentication mechanisms.

That said, if you need this value for DIGEST-MD5, you can store it in the object as an instance variable. If you reuse the responder for multiple authentication exchanges for different `digest-uri` values, you could set it before starting the exchange. I don't think it makes any sense if the value changes between the start of a subsequent authentication, and then responding to a new challenge by the server because it wants you to type in the password again.

> > This message on the IETF SASL mailing list might be interesting, by the way.
>
> Interesting indeed. But this is very forward-looking...
> Digest will be around for years, and the other common SASL mechanisms (CRAM-MD5, PLAIN) are more insecure.
>
> (from a bit of reading [2831bis](http://tools.ietf.org/html/draft-ietf-sasl-rfc2831bis), it seems to add even more complexity... let's hope we don't have to support it)

Sure, just a heads-up I received on the XMPP Standards list.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_36"></a>|@pitrou commented|
|-|-|

Replying to [ralphm](#note_35):
> Apparently you are talking about 'Subsequent Authentication'. Could you please use the terminology as used in the specification, because otherwise I get very confused.

Sure.

> That said, if you need this value for DIGEST-MD5, you can store it in the object as an instance variable. If you reuse the responder for multiple authentication exchanges for different `digest-uri` values, you could set it before starting the exchange.

Ok. It doesn't seem cleaner to me, but you can make the change if you want.
I suppose that the idea is to pass the digest-uri as an `__init__` parameter, but be able to modify it later.

I'm all for docstrings cleanup as well :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun set owner to @pitrou|
|-|-|

Some general points:
  * using unicode throughout would seem to be a better idea than using str throughout (or allowing either str or unicode, as is the case in a few places).  The SASL RFC talks extensively about unicode and code points.  Beyond that, for general str/unicode practices, pushing encoding to the i/o boundary just makes sense.  Since SASL keeps this pretty complex (eg the latin-1 default, as codified in BaseDigestMechanism.specialEncode), it may make sense to accept ```str``` instead of ```unicode``` as input (although this is something we should do regretfully), but any particular interface should take or produce one or the other, not both.
  * nothing uses IAuthzID.  it should be omitted until it offers some functionality.  this probably means ISASLCredentials isn't necessary yet, either.  It seems the idea of an authorization identifier from SASL may stretch cred a bit.  Previously this was a concept which was specifically supported in Twisted, but the complexity and lack of use cases led it to be dropped.  Now fitting it back into cred is going to take some thinking.  A lot of benefit can still be derived from SASL without implementing this portion of it, though.
  
Some trivial things:
  * twisted/cred/sasl.py
    * should start with a test-case-name declaration.  emacs and buildbot both use this to know what tests are relevant for a file.
    * a reference to rfc 4422 in the module docstring would probably help people know what's going on here
    * zope.interface import belongs above twisted import.  imports should proceed downwards in dependency order.
    * some lines exceed 80 cols
    * the first sentence in the docstring for ISASLResponder is a little unclear.  Presumably "auth server" isn't intended to refer to a dedicated authorization?/authentication? server, but a SASL challenger speaking via a SASL-enabled protocol to the responder.
    * ISASLResnder.getInitialResponse doesn't document its uri parameter.  In getResponse, it is documented, but not clearly enough for me to know what value I might want to pass to it, as a protocol implementor, nor what to do with it, as a sasl mechanism implementor.
    * ISASLChallenger.getSuccessfulChallenge has `self' in its parameter list
  * twisted/cred/plain.py
    * some lines exceed 80 cols
    * comment in PlainCredentials class scope suggests a bug somewhere else in cred.  Has a ticket been filed for this?
    * PlainCredentials looks a lot like UsernamePassword.  Without IAuthzID, does it still make sense to have it?
    * SASLPlainResponder should accept one or the other of str and unicode, not both.
    * Class docstring for SASLPlainResponder should document its attributes
    * SASLPlainChallenger's methods are all missing docstrings
  * twisted/test/test_plain.py
    * some lines exceed 80 cols
    * module docstring missing
    * class docstrings missing
    * test docstrings should all be complete sentences and describe the functionality the test is covering.
    * tests for exceptional cases should be catching the most specific exception type they can, rather than just LoginError
  * twisted/cred/digest.py
    * some lines exceed 80 cols
    * import order between zope/twisted.  stdlib imports belong before either.
    * secureRandom may belong in a different module (and see the top of twisted/names/dns.py and twisted/conch/ssh/common.py)
    * preHA1 branch of calcHA1 has no test coverage at all
    * the branch in BaseDigestResponder.getResponse which raises UnexpectedFinalChallenge has no test coverage
    * no test coverage for stale challenges
    * SASLDigestResponder.getInitialResponse has no test coverage
    * most/all error checking code in BaseDigestChallenger.processResponse is untested
    * getSuccessfulChallenge method of the same class has untested codepath for missing qop handling
    * acceptNonce method of the same class has no test coverage for unparsable nonce/timestamps
  * twisted/test/test_digest.py
    * some lines exceed 80 cols
    * import order between twisted/stdlib
    * no module docstring
    * missing class docstrings
    * test docstrings similarly incomplete
    * the url referenced aboved quotedWWWChallenge is dead and I'm not sure what the ```#p6``` was pointed at.

I haven't looked at the concrete patch yet, only the changes in the branch now.  Thanks for all the work you've put into this, sorry it's taken so long to get another review of it done.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_39"></a>|@pitrou commented|
|-|-|

Thanks a lot for the reply!

I'm gonna post a first patch for all the requested small fixes (docstrings, line wrapping, etc.).
Here are some answers to the remaining points:

Replying to [exarkun](#note_38):
> Some general points:
>   * using unicode throughout would seem to be a better idea than using str throughout (or allowing either str or unicode, as is the case in a few places).  The SASL RFC talks extensively about unicode and code points.  Beyond that, for general str/unicode practices, pushing encoding to the i/o boundary just makes sense.

Agreed, it will also make the code simpler.

> Since SASL keeps this pretty complex (eg the latin-1 default, as codified in BaseDigestMechanism.specialEncode), it may make sense to accept ```str``` instead of ```unicode``` as input (although this is something we should do regretfully), but any particular interface should take or produce one or the other, not both.

Ok.

>   * nothing uses IAuthzID.  it should be omitted until it offers some functionality.

Well, it does offer a piece of functionality: the ``authzid`` attribute, which application code is free to retrieve and use in its own manner. Although it isn't used anywhere in the Twisted code, it can be useful and I'm not sure removing it buys us anything.

> Some trivial things:
>   * twisted/cred/sasl.py
>     * should start with a test-case-name declaration.  emacs and buildbot both use this to know what tests are relevant for a file.

Well, there is no test-case-name to give since the file is only declaring interfaces and there is no test case I can think of :)
Should I put an empty test-case-name declaration instead?

>     * comment in PlainCredentials class scope suggests a bug somewhere else in cred.  Has a ticket been filed for this?

Yes, and fixed: [#2523](https://github.com/twisted/twisted/issues/2523)

>     * SASLPlainChallenger's methods are all missing docstrings

Mmmh, they are implementations of the methods documented in the ISASLChallenger interface. Do I have to write specific docstrings anyway?

>     * secureRandom may belong in a different module (and see the top of twisted/names/dns.py and twisted/conch/ssh/common.py)

Ok. Which module do you suggest?

Thanks

Antoine.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_40"></a>|@pitrou commented|
|-|-|

Just uploaded a patch for a lot of small fixes: docstrings, module import order, checking specific errors instead of twisted.cred.error.LoginFailed, 80-col line wrapping, removing references to solved twisted.cred.portal bug, replacing dead URL.
Can someone commit it to the branch?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun commented|
|-|-|

Replying to [antoine](#note_39):
> Thanks a lot for the reply!
>
> I'm gonna post a first patch for all the requested small fixes (docstrings, line wrapping, etc.).
> Here are some answers to the remaining points:

Cool, I'll apply that to the branch.

>
> Replying to [exarkun](#note_38):
> > Some general points:
> [snip]
>
> >   * nothing uses IAuthzID.  it should be omitted until it offers some functionality.
>
> Well, it does offer a piece of functionality: the ``authzid`` attribute, which application code is free to retrieve and use in its own manner. Although it isn't used anywhere in the Twisted code, it can be useful and I'm not sure removing it buys us anything.

The primary thing I think is gained by removing it (temporarily) is simplicity.  The branch adds a bunch of functionality.  We don't need to add everything simultaneously.  I'm not even suggesting we wait until after the next release to add authorization identifier support, just until after the rest of this functionality is merged.  It'll be easier to think about potential issues if the feature can be thought about independently of the rest of sasl.

>
> > Some trivial things:
> >   * twisted/cred/sasl.py
> >     * should start with a test-case-name declaration.  emacs and buildbot both use this to know what tests are relevant for a file.
>
> Well, there is no test-case-name to give since the file is only declaring interfaces and there is no test case I can think of :)
> Should I put an empty test-case-name declaration instead?

For this kind of file, it's usually a good idea to put a package name as a test-case-name.
Unfortunately there is no test package dedicated to cred (maybe there should be - separate issue).  The value can be a comma separated list of names, though.  Putting the name of one
or more cred-related test modules there would be good.

>
> >     * comment in PlainCredentials class scope suggests a bug somewhere else in cred.  Has a ticket been filed for this?
>
> Yes, and fixed: [#2523](https://github.com/twisted/twisted/issues/2523)

Awesome :)

>
> >     * SASLPlainChallenger's methods are all missing docstrings
>
> Mmmh, they are implementations of the methods documented in the ISASLChallenger interface. Do I have to write specific docstrings anyway?

Yep :)  Docstrings on the interface should describe the reason the method exists, what it is expected to do, etc.  Docstrings on the implementation should describe whatever is special about that particular implementation (if there isn't anything special, that's a hint that the interface might not be appropriate).

>
> >     * secureRandom may belong in a different module (and see the top of twisted/names/dns.py and twisted/conch/ssh/common.py)
>
> Ok. Which module do you suggest?
>

Probably somewhere in twisted.python...  it would make sense to have a separate ticket factoring duplication in conch and names into a single module, and then use that module for this ticket.

Thanks :)

Jean-Paul

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_42"></a>|@pitrou commented|
|-|-|

The sasl-unicode-strict patch assumes unicode parameters (for username/password) at SASL interface boundaries - compared to the previous code which allowed either str or unicode. In the process, I've fixed a bug related to processing of non-ascii responses, and added tests for it.

By the way, a problem is that t.cred.credentials declares credentials attributes as str rather than unicode (see e.g. IHashedUsernamePassword).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_43"></a>|@therve commented|
|-|-|

(In [21376]) Merge secure-random-2685-3

Authors: antoine, therve
Reviewers: exarkun, radix
Fixes [#2685](https://github.com/twisted/twisted/issues/2685)
Refs [#2015](https://github.com/twisted/twisted/issues/2015)

Centralize a secure random implementation in twisted.python.randbytes (relying
on different backends, like os.urandom of PyCrypto randpool), to produce random
bytes in a secure fashion. Also, replace custom work in twisted.names.dns and
twisted.ssh.common.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_44"></a>|@therve removed owner|
|-|-|

This ready to review in generic-sasl-2015-2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_45"></a>|@exarkun set owner to @pitrou|
|-|-|

IAuthzID still doesn't make sense to me.  I thought about the functionality and talked to some people on IRC about it and I think I understand how RFC 2222 _authorization identifier_ support should work in cred.  I'll try to explain it here, but first let me try to explain what's wrong with the IAuthzID code which is currently in the branch.

If a protocol implementation wants to support SASL authzid, then it will need to create a credential object which provides IAuthzID.  Since IAuthzID only has the authzid attribute, the credential object must also provide some other credential interface (such as IUsernamePassword).  Any protocol code can be changed to do this (assuming the underlying protocol actually supports SASL/authzid).  It then gives this credential object to cred via portal.login and waits for an avatar to come back.    Any credentials checker which claims to implement an interface the credential object provides will be invoked to verify it.  Since IAuthzID isn't expressive enough to authenticate anything, checkers won't include it in their credentialInterfaces list.  Instead, they'll include something like IUsernamePassword.  An IAuthzID-aware checker will get the credential object, check to see if it provides IAuthzID in addition to the main credential interface, and then take some special action if it does (I'll get back to this).  However, a checker which is not aware of IAuthzID can also be invoked this way.  It will ignore IAuthzID and indicate some other value for the avatar identifier (which is similar-ish to authzid).  This will probably result in the user being logged in with a different identity than they expected.  It might also result in the login failing (if the authentication identifier is not a valid authorization identifier/avatar identifier) or in an even worse case, if the authentication identifier is a valid authorization identifier/avatar identifier _for a different user_.  Clearly, an undesirable outcome.  Avoiding this requires careful coordination between protocol implementations and portal configurations.  Worse, if we add authzid support to all of Twisted's protocol implementations, then no existing checkers will be suitable for use with them any more.  Getting back to the "special action" I mentioned above: if the checker is aware of IAuthzID, it will need to determine somehow whether the given value is allowed for the given authentication identifier (assuming the password or other credential information is verified, which shouldn't pose any new problems).  In order to do this, it needs to do authorization, a task which is currently left up to the realm.  Crossing responsibilities like this couples checkers tightly to realms, something cred is designed to avoid.

I hope that made some sense.  Here's what I propose as an alternative:

A protocol implementation which intends to support SASL and authzid should, when presented with an authentication attempt, first attempt a login to its portal with the credentials it has been presented _without_ including the authzid information.  It should attempt this login using a special interface, not the usual protocol interface (eg, not _IMailbox_ for pop3).  The special interface represents the ability to select the authorization user explicitly.  Since the login is attempted only with the normal credentials interface (eg, IUsernamePassword), all existing checkers will work correctly with it.  A realm which supports the notion of authzid will have to support this special interface and return an appropriate avatar which implements it.  The interface has only one method, which allows an attempt to be made to get an avatar for a different interface and different authorization user.  The realm can construct this avatar such that the attempt will succeed only for the authorization users as whom the authentication user is allowed to masquerade.  As a straw-man, here is what some protocol code might look like:

```python
    def command_LOGIN(self, username, password, authzid):
        d = self.portal.login(UsernamePassword(username, password), self, IRealm)
        def cbLoggedIn((interface, avatar, logout)):
            assert interface is IRealm
            self.logout = logout
            return avatar.requestAvatar(authzid, self, IMailbox)
        d.addCallback(cbLoggedIn)
        def cbSwitched((interface, avatar, logout)):
            assert interface is IMailbox
            self.logout()
            self.avatar = avatar
            self.logout = logout
        d.addCallback(cbSwitched)
        return d
```

Using ```IRealm``` is probably too cute, but I couldn't resist for this comment. ;)


Glyph and I had a long discussion about this approach and he pointed out that it fails for the case where the relationship between the authzid and the authentication identifier is known to the checker, rather than the relationship between the the checker-translated authorization identifier which corresponds to the authzid and the avatar identifier/authorization identifier.  That might be fixable with a realm wrapper which knows about the checker or the database the checker users and intercepts login attempts which require that information (and passes the rest through to the original realm).  This requires some custom realm code, but not much.

Glyph may post a comment arguing a different point of view later on.  Whether he does or not, given the complexity represented solely in my above comment, I think the IAuthzID stuff should come out of this branch and be implemented later separately.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_47"></a>|@therve removed owner|
|-|-|

I think I removed everything regarding authzid. Back to review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_48"></a>|@exarkun set owner to @therve|
|-|-|

Some other stuff which apparently I couldn't see while ```IAuthzID``` was present:

  * ```parseChallenge``` and ```parseResponse``` are both for dealing with untrusted input from the network; they should at least have their exceptions well-defined (and tested)
  * The code path through ```calcHA1``` which runs if ```pszAuthzID``` is true isn't tested
  * ```BaseDigestMechanism.__doc__``` has a typo - _mechanim_
  * ```BaseDigestMechanism.__init__``` still accepts ```str``` or ```unicode``` for some reason
  * ```BaseDigestMechanism.decode``` seems pointless
  * ```BaseDigestMechanism._guessAlgorithm``` should be public or ```BaseDigestMechanism``` should be private, since it won't actually work unless a subclass overrides that method.
  * The code path through ```BaseDigestResponder.getResponse``` which runs if ```f['stale']``` is true isn't tested
  * ```SASLDigestResponder.getInitialResponse``` is untested
  * ```SASLDigestResponder.getResponse``` violates the principle of _Liskov Substitutability_ (apologies for saying that) - probably the base implementation shouldn't be called ```getResponse``` (maybe this doesn't matter - seems less important the more I think about it)
  * The code path through ```BaseDigestChallenger.getSuccessfulChallenge```
  * Most of the test strings at the top of ```test_digest.py``` don't need to be global since only one test method uses each of them
  * Various ```XXX``` comments should be addressed

I started reading the encoding rules in the RFC but it's too late and they're too crazy.  It kinda looks like everything is okay in that department, although I don't really like the repeated use of the string "utf-8" (a symbolic name for that would be nice).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_49"></a>|@therve commented|
|-|-|

(In [21928]) Stuff

Refs [#2015](https://github.com/twisted/twisted/issues/2015)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_50"></a>|@therve commented|
|-|-|

(In [21929]) Tests.

Refs [#2015](https://github.com/twisted/twisted/issues/2015)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_51"></a>|@therve commented|
|-|-|

(In [21930]) Add test to parseResponse errors, change getResponse

Refs [#2015](https://github.com/twisted/twisted/issues/2015)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_52"></a>|@khorn commented|
|-|-|

I would really like to see this ticket move forward.

What remains to be done?
What can I do to help?

As a first step, I reviewed exarkun's list of TODO's above (from 2007-11-13 03:29:08+00:00) and compared it with the 3 changesets submitted by therve.  Here's what I came up with:

    * [DONE] ```parseChallenge``` and ```parseResponse``` are both for dealing with untrusted input from the network; they should at least have their exceptions well-defined (and tested)
    * [DONE] The code path through ```calcHA1``` which runs if ```pszAuthzID``` is true isn't tested
    * [Done] ```BaseDigestMechanism.__doc__``` has a typo - mechanim
    * ```BaseDigestMechanism.__init__``` still accepts str or unicode for some reason
    * [DONE] ```BaseDigestMechanism.decode``` seems pointless
    * [DONE] ```BaseDigestMechanism._guessAlgorithm``` should be public or ```BaseDigestMechanism``` should be private, since it won't actually work unless a subclass overrides that method.
    * [SOME?] The code path through ```BaseDigestResponder.getResponse``` which runs if ```f['stale']``` is true isn't tested
    * [SOME?] ```SASLDigestResponder.getInitialResponse``` is untested
    * [DONE?] ```SASLDigestResponder.getResponse``` violates the principle of Liskov Substitutability (apologies for saying that) - probably the base implementation shouldn't be called ```getResponse``` (maybe this doesn't matter - seems less important the more I think about it)
    * [SOME?] The code path through ```BaseDigestChallenger.getSuccessfulChallenge```
    * [DONE] Most of the test strings at the top of ```test_digest.py``` don't need to be global since only one test method uses each of them
    * Various XXX comments should be addressed 

NOTE: my opinions on the 'done-ness' of any of these items should be viewed as highly suspect...please provide your opinions!

The only issue that hasn't really been touched is ```BaseDigestMechanism.__init__``` accepting unicode and assuming UTF-8.  The relevant RFC's are mostly silent on this, and appear to assume 

[This post](http://stackoverflow.com/questions/702629/utf-8-characters-mangled-in-http-basic-auth-username/703341#703341) on Stack Overflow has some good info on how browser deal with this issue for Basic HTTP Auth.

To me the whole issue is unclear and I think Twisted should "refuse the temptation to guess".

I also think this ticket should just cover what's in the branch now (and appropriate changes, further tests, etc.) with further tickets covering the porting of existing Twisted protocol implementations (the concrete-patch stuff) to use this SASL API.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_54"></a>|@khorn removed owner|
|-|-|

After reading a *lot* of RFCs and opinions on the issue of `BaseDigestMechanism.__init__` accepting unicode, and assuming UTF-8, it appears that the answer is in [RFC2831](http://www.faqs.org/rfcs/rfc2831.html).

The SASL Digest mechanism accepts a "charset" directive, the only valid value of which is the string "utf-8".  If the directive is present, the encoding is assumed to be (you guessed it) UTF-8.  If the directive is *not* present the encoding is assumed to be ISO-8859-1.  The assumption is that in HTTP Digest authentication the directive will not be present (since it isn't defined), so the encoding will be ISO-8859-1, which it (probably) should be.  When using the SASL Digest mechanism, however, you can use either encoding, indicated by the presence or absence of this directive.

So with all this in mind, it looks as though the original code is correct, or at least as correct as it can reasonably get, considering the confusion in the RFCs regarding HTTP Basic and Digest authentication.

Sending this to the review queue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_55"></a>|@exarkun set owner to @khorn|
|-|-|

Thanks for the investigation on this, khorn.

Itamar and I took a brief look at this branch and came up with a couple high level issues that need to be addressed here.

  1. First, trunk has moved on a bit since this was last worked on.  Twisted Cred actually has a bunch of digest auth code in it now which this branch is duplicating.  So, hopefully there's a bunch of icky crypto code that can be dropped from this branch entirely and replaced with the APIs in trunk already.
  1. Next, some documentation is sorely needed.  It's not clear what this branch is actually adding, nor how one would go about using this in a protocol implementation, or how one would extend it with another sasl mechanism.  Someone who knows what's going on here should describe how these new pieces fit together with the existing pieces in Twisted (in the form of a new cred howto, I think).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_56"></a>|@khorn commented|
|-|-|

Replying to [exarkun](#note_55):
> Thanks for the investigation on this, khorn.
>
> Itamar and I took a brief look at this branch and came up with a couple high level issues that need to be addressed here.
>
>   1. First, trunk has moved on a bit since this was last worked on.  Twisted Cred actually has a bunch of digest auth code in it now which this branch is duplicating.  So, hopefully there's a bunch of icky crypto code that can be dropped from this branch entirely and replaced with the APIs in trunk already.

Ah, I see.  I'll see if I can come up with a patch against a more recent version of trunk.  It looks like there is a *lot* more digest stuff in this ticket than there is in the trunk, though.

>   2. Next, some documentation is sorely needed.  It's not clear what this branch is actually adding, nor how one would go about using this in a protocol implementation, or how one would extend it with another sasl mechanism.  Someone who knows what's going on here should describe how these new pieces fit together with the existing pieces in Twisted (in the form of a new cred howto, I think).

Agreed that some documentation would probably help.  Here's a few notes (off the top of my head, so grain of salt, etc.):

As for what this branch is trying to do, the idea is to have all the code necessary to add SASL to your protocol in *one place*.  Right now there are bits of SASL all over Twisted (imap, sip, xmpp), and this had led to unnecessary duplication of code and slight differences in the implementations (for example, as of this week, XMPP supports SASL ANONYMOUS, but nothing else does AFAIK).  It also means more code to maintain.

According to my current understanding of the code (which is admittedly not perfect), the way you would use it in a protocol would be similar to the way PLAINCredentials/PLAINAuthenticator are used in the IMAP code.  I think there is some room for improvement here, but for now I think we should just focus on getting everything in one place.

To implement another SASL mechanism, you would implement ISASLResponder on the client end, and ISASLChallenger on the server end, then have your protocol use those objects to manage the SASL conversation.  Keep in mind that the protocols will still need to "wrap" the challenges and responses in their own protocol-specific ways.

Now this ticket is positively ancient, so it's obvious there needs to be some work done in catching it up, but I think the end result will be worth it.

So here's where I see this going (probably not all on this ticket):

 1. Get this code caught up to trunk, removing duplication of digest code that is now in Cred.
 1. Port protocols with existing SASL implementations (IMAP, SIP, XMPP) to use this one. (separate tickets, I'm thinking)
 1. Maybe come up with an higher level API that would make it easier to "drop in" SASL into a protocol.  Obviously each protocol uses SASL somewhat differently, but I think some work could be done here.  Maybe some kind of manager object which handles all of the SASL mechanisms for that protocol.  Maybe some new interfaces (ISASLClientProtocol, ISASLServerProtocol).  I don't know. This is down the road a bit.
 1. Obviously, docs on how this all fits together, and how it fits together with Cred.

But the first step, IMO, is to get things consolidated into one place.  I'll see about getting things updated to something more current, and see if I can't get rid of the digest duplication.  And I'll see if I can't come up with some simple docs on how to use this in a protocol.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_58"></a>|@exarkun commented|
|-|-|

Cool.  I'm glad you're still interested in this area.  It'll be a wonderful improvement to cred and the protocol implementations in Twisted when it finally lands. :)

I'll just point out that I think point 4 from your list above should probably be somewhat high priority.  Catching up to trunk probably makes sense to do first, but I'd probably tackle docs immediately after that.  This should help shake out any problems with the approach and will make it easier for more people to get involved.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="whitelynx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_60"></a>|whitelynx commented|
|-|-|

It may also be good to look at patching [https://github.com/thisismedium/python-sasl] to work on Python 2.4 and up, and use that for our SASL implementation.
